design/construct-tree.md

```diff
@@ -0,0 +1,343 @@
+    4. **Assets** - Assets information will be included in the construct tree. Assets represent actions that the CDK *will* take ahead of stack deployments (i.e. S3 assets are zipped and uploaded directly, Docker images are uploaded to ECR). The asset metadata in the tree is the S3 key that an asset *would* have
+3. **Local** - The construct tree is produced at synthesis time. Connectivity to AWS and CloudFormation should not be required to produce a tree view of a CDK application.
+
+## Approach
+
+At a high-level, we will render the `construct tree` through a construct tree data model and construct metadata interface that constructs will implement. This section will describe the components of the JSON structure that will represent the construct-tree:
+
+### Construct Tree Data Model
+
+A feature will be added to the CDK core module so that any CDK application will produce an artifact called `tree.json` during calls to `cdk synth` and include it as an output to the cloud assembly in `cdk.out`. Developers will be able to opt-out if this functionality is not desired or needed through a flag in the CLI or set it as a preference.
+
+The construct tree will be a list of paths that are indexed into a flat map of constructs. The file structure itself represents the tree. This information will need to include the following to start rendering a flat map of constructs:
+
+### Construct properties
+
+|Property    |Type    |Required |Description  |
+|---  |---  |---  |---  |
+|id |string |Required |id of the construct within the current scope |
+|path |string |Required |full, absolute path of the construct within the tree |
+|children |Array<Path>  |Not Required |All direct children of this construct. Array of the absolute paths of the constructs. Will be used to walk entire list of constructs |
+|metadata |Array<Metadata>  |Not Required |Metadata describing all constructs/resources that are encapsulated by the construct |
+
+### Metadata Properties
+
+The following metadata properties will be included by the construct that produces the `tree.json` output.
+
+|Property    |Type    |Required    |Description    |
+|---    |---    |---    |---  |
+|sourceLocation    |string    |Required    |location in source code where the construct is defined    |
+|resourceType |string |Required |type of resource (schema / classification terminology TBD) |
+|description    |string |Not Required |description of the construct |
+|renderedView    |Array<Metadata> |Not Required    |constructs can fill in arbitrary. ClouFormation agnostic. CFN constructs could pre-populate it with the output of `this.toCloudFormation()` |
```

Comment:
I'm not sure what this is? What should non-CFN Resource constructs populate here, if they don't translate into text, like a binary?

Further, I don't see a requirement that warrants this. Do we need it?